6_predict/src/plot_diagnostics.R

```diff
@@ -261,6 +261,66 @@ plot_pred_obs <- function(df_pred_obs, metric, region, out_dir,
 }
 
 
+make_residual_map <- function(df_pred_obs, sites_g2_sf, metric, pred_gage_ids, region, out_dir,
+                              from_predict = FALSE, model_wf = NULL, pred_data = NULL){
+  #' @description this function creates maps of residuals (obs - predicted) and a
+  #' variogram plot
+  #' 
+  #' @param df_pred_obs is a data frame with observed (obs) and predicted (.pred) from 
+  #' the output of  predict_test_data function. if left null and from_predict is true, 
+  #' it will predict it from the model_wf (copied these lines from the plot_pred_obs function)
+  #' @param sites_g2_sf  gagesii spatial ata object with LAT LON and ID
+  #' @param metric  metric that is being predicted
+  #' @param pred_gage_ids vector of gage ids that correspond to each row of the df_pred_obs
+  #' @param region  region of the model/prediction
+  #' @param directory where output figures are saved
+  #' 
+
+  if(from_predict){
+    #predict from provided workflow and data
+    df_pred_obs <- predict(model_wf, new_data = pred_data, type = 'numeric') %>%
+      mutate(obs = pred_data[[metric]])
+  }
+  
+  lat_lons<- sites_g2_sf %>%
+    filter(ID %in% pred_gage_ids) %>%
+    select(ID, LAT, LON)
+  
+  df<- bind_cols(lat_lons,df_pred_obs) %>%
+    mutate(resid = obs - .pred)
+  
+  states <- map_data("state")
+  
+  p1<-ggplot(states, aes(x=long, y=lat, group=group)) +
+    geom_polygon(fill="white", colour="gray") +
+    geom_sf(data = df, inherit.aes = FALSE, 
+            aes(color = .data[["resid"]]), 
+            size = 0.75)+
+    scale_color_scico(palette = 'batlow')+
+    theme(legend.position="bottom",
+          legend.key.size=unit(.5,'cm'))+
+    xlab('Longitude') + 
+    ylab('Latitude')+
+    ggtitle(paste("metric= ",metric,"    region = ", region ))
+  
+  
+  p2 <- plot(variogram(resid~1, df))
```

Comment:
What are the distance units? km? I think we should increase the xlimit to 100 km and increase the number of lags to compute to 50 or so

There are a few outlier residuals that result in high short-distance semivariance. We can try plotting a 3rd panel with a robust (to outlier point-pairs) semivariance. There is a function to do that within gstat but I don't remember off the top of my head.

Reply:
cressie = TRUE for robust semi-variogram

Reply:
I actually don't really know what I am doing with the variogram.  I think the limit and number of lags are set with cutoff and width, respectively? I'm not sure what the units are though - the x and y are in lat lon, so I guess they are in degrees? Should I project it first so they have units of distance?

Reply:
Based on the documentation, you don't need to use a projected coordinate system. You can set `projected = FALSE` and distances will be computed in km units.
Are we currently using a projected or unprojected coordinate system?

Yes, the cutoff will set xmax and width sets the increments between points. So I'd suggest using 100 km for cutoff and 2 km for width.